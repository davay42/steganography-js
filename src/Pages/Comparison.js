import React, { Component } from "react";
import Noty from "noty";
import LangContext from "../context/LangContext";
import CanvasSection from "../compontens/CanvasSection/CanvasSection";
import ColorSelector from "../compontens/ColorSelector/ColorSelector";
import ProgressBar from "../compontens/ProgressBar/ProgressBar";

export default class Comparison extends Component {
  constructor(props) {
    super(props);
    this.state = {
      processActive: false,
      progress: 0,
      firstFile: "",
      secondFile: "",
    };
    this.differ = null;
  }

  firstFileLoaded(bool, name) {
    this.setState({ firstFile: true });
  }

  secondFileLoaded(bool, name) {
    this.setState({ secondFile: true });
  }

  onDifferMessage(e, transferList) {
    const t = this.context;

    if (e.data.progressBar) {
      this.setState({ progress: e.data.progressBar });
    }

    if (e.data.error) {
      this.setState({ processActive: false, progress: 0 });

      new Noty({
        theme: "nest",
        type: "error",
        layout: "topRight",
        text: t("notification:error_noty", { message: e.data.error }),
        timeout: 2000,
      }).show();
    }

    if (e.data.done) {
      this.setState({ progress: 100 });

      let fileName = "diff_stegojs_" + (Math.random() * 10000).toFixed();

      // Image mode result
      // Create a TypedArray from the transferred ArrayBuffer
      let typed = new Uint8ClampedArray(e.data.result.buffer);
      // Create an ImageData object to draw to the canvas
      let resultImage = new ImageData(
        typed,
        e.data.result.width,
        e.data.result.height,
      );
      // Draw the new image
      this._resultCanvas.updateImage(
        resultImage,
        fileName + ".bmp",
      );
      this._resultCanvas.scale("in");

      new Noty({
        theme: "nest",
        type: "success",
        layout: "topRight",
        text: t("notification:diffing_finished", { time: e.data.done }),
        timeout: 5000,
      }).show();

      // Terminate worker
      this.differ.terminate();
      this.differ = null;

      setTimeout(() => {
        this.setState({ processActive: false, progress: 0 });
      }, 1000);
    }
  }

  diff() {
    // Clear the previous result
    this._resultCanvas.resetState();

    // Get the first image
    let canvas = document.getElementById("first-image-canvas");
    let ctx = canvas.getContext("2d");
    let imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
    let firstImage = {
      buffer: imageData.data.buffer,
      width: canvas.width,
      height: canvas.height,
    };

    // Get the second image
    let canvas2 = document.getElementById("second-image-canvas");
    let ctx2 = canvas2.getContext("2d");
    let imageData2 = ctx2.getImageData(0, 0, canvas2.width, canvas2.height);
    let secondImage = {
      buffer: imageData2.data.buffer,
      width: canvas2.width,
      height: canvas2.height,
    };

    // Get the selected highlight color
    const highlight = this._colorSelector.getColor();

    // Merge params
    const payload = {
      process: "diff",
      mode: "DEFAULT", // TODO: create modes
      diffColor: highlight,
      first: firstImage,
      second: secondImage,
    };
    const transferList = [firstImage.buffer, secondImage.buffer];

    this._resultCanvas.scale("out");

    this.setState({ processActive: true });
    this.differ = new Worker(
      `${process.env.PUBLIC_URL}/workers/differ.worker.js`,
    );
    // Set handler
    this.differ.onmessage = this.onDifferMessage.bind(this);
    // Start process
    this.differ.postMessage(payload, transferList);
  }

  render() {
    const t = this.context;
    return (
      <div className="App-content pad-top">
        <ProgressBar
          active={this.state.processActive}
          progress={this.state.progress}
        />
        <div id="grid-wrapper-vertical">
          <div
            className="grid-element"
            id="first-section"
            style={{ gridColumnStart: 1 }}
          >
            <h5 className="section-title">{t("comparison:first_image")}</h5>
            <CanvasSection
              ref={(ref) => this._firstImage = ref}
              id="first-image"
              onFileLoaded={(bool, name) => this.firstFileLoaded(bool, name)}
              isAProcessActive={this.state.processActive}
              hideCanvas
            />
          </div>
          <div
            className="grid-element"
            id="second-section"
            style={{ gridColumnStart: 1 }}
          >
            <h5 className="section-title">{t("comparison:second_image")}</h5>
            <CanvasSection
              ref={(ref) => this._secondImage = ref}
              id="second-image"
              sourceFileLoaded="true"
              onFileLoaded={(bool, name) => this.secondFileLoaded(bool, name)}
              isAProcessActive={this.state.processActive}
              hideCanvas
            />
          </div>
          <div
            className="grid-element"
            id="highlight-section"
            style={{ gridColumnStart: 1 }}
          >
            <h5 className="section-title">{t("comparison:diff_color")}</h5>
            <ColorSelector ref={(ref) => this._colorSelector = ref} />
          </div>

          <div
            className="grid-element"
            id="result-section"
            style={{ gridRow: "1 / span 4", gridColumnStart: 2 }}
          >
            <h5 className="section-title">{t("common:difference")}</h5>
            <CanvasSection
              style={{ width: 2 }}
              ref={(ref) => this._resultCanvas = ref}
              id="result-image"
              disableInput
            />
            <div className="section-actions">
              <button
                className="waves-effect waves-light btn grey darken-4"
                disabled={!this.state.firstFile || !this.state.secondFile ||
                  this.state.processActive}
                onClick={() => this.diff()}
                style={{ margin: "auto" }}
              >
                {t("common:compare")}
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Comparison.contextType = LangContext;
