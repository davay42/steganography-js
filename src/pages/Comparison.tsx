import React, { Component, createRef } from "react";
import Noty from "noty";
import LangContext from "../context/LangContext";
import CanvasSection from "../components/CanvasSection/CanvasSection";
import ColorSelector from "../components/ColorSelector/ColorSelector";
import ProgressBar from "../components/ProgressBar/ProgressBar";

interface State {
    processActive: boolean,
    progress: number,
    firstFile: boolean,
    secondFile: boolean,
}

export default class Comparison extends Component<{}, State> {
    _resultCanvas = createRef<CanvasSection>();
    _colorSelector = createRef<ColorSelector>();
    _firstImage = createRef<CanvasSection>();
    _secondImage = createRef<CanvasSection>();

    state = {
        processActive: false,
        progress: 0,
        firstFile: false,
        secondFile: false,
    };

    differ: Worker | null = null;

    firstFileLoaded = (bool: boolean, name: string) => {
        this.setState({ firstFile: bool });
    }

    secondFileLoaded = (bool: boolean, name: string) => {
        this.setState({ secondFile: bool });
    }

    onDifferMessage = (e: MessageEvent) => {
        const t = this.context;

        if (e.data.progressBar) {
            this.setState({ progress: e.data.progressBar });
        }

        if (e.data.error) {
            this.setState({ processActive: false, progress: 0 });

            new Noty({
                theme: "nest",
                type: "error",
                layout: "topRight",
                text: t("notification:error_noty", { message: e.data.error }),
                timeout: 2000,
            }).show();
        }

        if (e.data.done) {
            this.setState({ progress: 100 });

            const fileName = "diff_stegojs_" + (Math.random() * 10000).toFixed();

            // Image mode result
            // Create a TypedArray from the transferred ArrayBuffer
            const typed = new Uint8ClampedArray(e.data.result.buffer);
            // Create an ImageData object to draw to the canvas
            const resultImage = new ImageData(
                typed,
                e.data.result.width,
                e.data.result.height,
            );
            // Draw the new image
            this._resultCanvas.current?.updateImage(
                resultImage,
                fileName + ".bmp",
            );
            this._resultCanvas.current?.scale("in");

            new Noty({
                theme: "nest",
                type: "success",
                layout: "topRight",
                text: t("notification:diffing_finished", { time: e.data.done }),
                timeout: 5000,
            }).show();

            if (this.differ) {
                this.differ.terminate();
                this.differ = null;
            }

            setTimeout(() => {
                this.setState({ processActive: false, progress: 0 });
            }, 1000);
        }
    }

    diff = () => {
        this._resultCanvas.current?.resetState();

        // Get the first image
        const canvas = document.getElementById("first-image-canvas") as HTMLCanvasElement;
        const ctx = canvas.getContext("2d")!;
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
        const firstImage: ImagePayload = {
            buffer: imageData.data.buffer,
            width: canvas.width,
            height: canvas.height,
        };

        // Get the second image
        const canvas2 = document.getElementById("second-image-canvas") as HTMLCanvasElement;
        const ctx2 = canvas2.getContext("2d")!;
        const imageData2 = ctx2.getImageData(0, 0, canvas2.width, canvas2.height);
        const secondImage: ImagePayload = {
            buffer: imageData2.data.buffer,
            width: canvas2.width,
            height: canvas2.height,
        };

        // Merge params
        const payload = {
            process: "diff",
            mode: "image",
            diffColor: this._colorSelector.current?.getColor(),
            first: firstImage,
            second: secondImage,
        };
        const transferList = [firstImage.buffer, secondImage.buffer];

        this._resultCanvas.current?.scale("out");

        this.setState({ processActive: true });
        this.differ = new Worker(`${process.env.PUBLIC_URL}/workers/differ.worker.js`);
        this.differ.onmessage = this.onDifferMessage;
        this.differ.postMessage(payload, transferList);
    }

    render() {
        return (
            <div className="App-content pad-top">
                <ProgressBar
                    active={this.state.processActive}
                    progress={this.state.progress}
                />
                <div id="grid-wrapper-vertical">
                    <div
                        className="grid-element"
                        id="first-section"
                        style={{ gridColumnStart: 1 }}
                    >
                        <h5 className="section-title">{this.context("comparison:first_image")}</h5>
                        <CanvasSection
                            ref={this._firstImage}
                            id="first-image"
                            onFileLoaded={(bool, name) => this.firstFileLoaded(bool, name)}
                            isAProcessActive={this.state.processActive}
                            hideCanvas
                        />
                    </div>
                    <div
                        className="grid-element"
                        id="second-section"
                        style={{ gridColumnStart: 1 }}
                    >
                        <h5 className="section-title">{this.context("comparison:second_image")}</h5>
                        <CanvasSection
                            ref={this._secondImage}
                            id="second-image"
                            sourceFileLoaded={true}
                            onFileLoaded={(bool, name) => this.secondFileLoaded(bool, name)}
                            isAProcessActive={this.state.processActive}
                            hideCanvas
                        />
                    </div>
                    <div
                        className="grid-element"
                        id="highlight-section"
                        style={{ gridColumnStart: 1 }}
                    >
                        <h5 className="section-title">{this.context("comparison:diff_color")}</h5>
                        <ColorSelector ref={this._colorSelector} />
                    </div>

                    <div
                        className="grid-element"
                        id="result-section"
                        style={{ gridRow: "1 / span 4", gridColumnStart: 2 }}
                    >
                        <h5 className="section-title">{this.context("common:difference")}</h5>
                        <CanvasSection
                            ref={this._resultCanvas}
                            id="result-image"
                            disableInput
                        />
                        <div className="section-actions">
                            <button
                                className="waves-effect waves-light btn grey darken-4"
                                disabled={!this.state.firstFile || !this.state.secondFile ||
                                    this.state.processActive}
                                onClick={this.diff}
                                style={{ margin: "auto" }}
                            >
                                {this.context("common:compare")}
                            </button>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

Comparison.contextType = LangContext;
