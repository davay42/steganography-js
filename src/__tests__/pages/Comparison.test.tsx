import React from 'react';
import renderer from 'react-test-renderer';

import Comparison from '../../pages/Comparison';

it('renders Comparison', () => {
    const tree = renderer
        .create(<Comparison />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});
