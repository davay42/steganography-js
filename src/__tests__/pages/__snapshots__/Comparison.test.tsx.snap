// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Comparison 1`] = `
<div
  className="App-content pad-top"
>
  <div
    className="App-progress"
    id=""
  >
    <div
      className="progress-indicator-wrapper"
    >
      <div
        className="grey darken-4 progress-indicator "
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
    <label
      className="progress-description "
    >
      <span>
        {progress_bar:perf_info}
      </span>
    </label>
  </div>
  <div
    id="grid-wrapper-vertical"
  >
    <div
      className="grid-element"
      id="first-section"
      style={
        Object {
          "gridColumnStart": 1,
        }
      }
    >
      <h5
        className="section-title"
      >
        {comparison:first_image}
      </h5>
      <div
        className="CanvasSelection"
      >
        <div>
          <label
            className="label"
            htmlFor="first-image-input"
          >
            {image_input:select_file}
          </label>
          <br />
          <input
            id="first-image-input"
            onChange={[Function]}
            type="file"
          />
        </div>
        <br />
        <div
          className="section-content scale-transition scale-initial"
          hidden={true}
        >
          <label
            className="label"
            htmlFor="first-image-canvas"
          >
            Image
          </label>
          <canvas
            id="first-image-canvas"
          />
        </div>
        <div
          className="section-actions secondary"
        />
        <div
          className="section-actions"
        />
      </div>
    </div>
    <div
      className="grid-element"
      id="second-section"
      style={
        Object {
          "gridColumnStart": 1,
        }
      }
    >
      <h5
        className="section-title"
      >
        {comparison:second_image}
      </h5>
      <div
        className="CanvasSelection"
      >
        <div>
          <label
            className="label"
            htmlFor="second-image-input"
          >
            {image_input:select_file}
          </label>
          <br />
          <input
            id="second-image-input"
            onChange={[Function]}
            type="file"
          />
        </div>
        <br />
        <div
          className="section-content scale-transition scale-initial"
          hidden={true}
        >
          <label
            className="label"
            htmlFor="second-image-canvas"
          >
            Image
          </label>
          <canvas
            id="second-image-canvas"
          />
        </div>
        <div
          className="section-actions secondary"
        />
        <div
          className="section-actions"
        />
      </div>
    </div>
    <div
      className="grid-element"
      id="highlight-section"
      style={
        Object {
          "gridColumnStart": 1,
        }
      }
    >
      <h5
        className="section-title"
      >
        {comparison:diff_color}
      </h5>
      <div
        className="color-selector"
      >
        {color_selector:select_color}
        :
        <input
          defaultValue="#651fff"
          onChange={[Function]}
          type="color"
        />
        <span
          className="color-selector-preview"
          style={
            Object {
              "backgroundColor": "rgba(101,31,255,255)",
            }
          }
        />
      </div>
    </div>
    <div
      className="grid-element"
      id="result-section"
      style={
        Object {
          "gridColumnStart": 2,
          "gridRow": "1 / span 4",
        }
      }
    >
      <h5
        className="section-title"
      >
        {common:difference}
      </h5>
      <div
        className="CanvasSelection"
      >
        <br />
        <div
          className="section-content scale-transition scale-initial"
        >
          <label
            className="label"
            htmlFor="result-image-canvas"
          >
            Image
          </label>
          <canvas
            id="result-image-canvas"
          />
        </div>
        <div
          className="section-actions secondary"
        />
        <div
          className="section-actions"
        />
      </div>
      <div
        className="section-actions"
      >
        <button
          className="waves-effect waves-light btn grey darken-4"
          disabled={true}
          onClick={[Function]}
          style={
            Object {
              "margin": "auto",
            }
          }
        >
          {common:compare}
        </button>
      </div>
    </div>
  </div>
</div>
`;
